### ISSUE TITLE
rework weight sharing between layers

### ISSUE THREAD

**Issue by baykaner** [fetchai/ledger#1475]
Current implementation is to have another constructor for FullyConnected that takes a pointer to the other shared Op; however, it would be much simpler to have sharing occur at a higher level: e.g. if nodes share their op_ptr.

**Comment by evsmithx** [comment]
Test from multihead attention
  //  // rebuild
  //  using DataType  = typename TypeParam::Type;
  //  auto layer2 = *(fetch::ml::utilities::BuildLayer<TypeParam, LayerType>(dsp2));
  //
  //  // test equality
  //  layer.SetInput("MultiheadAttention_Query", query_data);
  //  layer.SetInput("MultiheadAttention_Key", key_data);
  //  layer.SetInput("MultiheadAttention_Value", value_data);
  //  prediction = layer.Evaluate(output_name, true);
  //
  //  layer2.SetInput("MultiheadAttention_Query", query_data);
  //  layer2.SetInput("MultiheadAttention_Key", key_data);
  //  layer2.SetInput("MultiheadAttention_Value", value_data);
  //  TypeParam prediction2 = layer2.Evaluate(output_name, true);
  //
  //  ASSERT_TRUE(prediction.AllClose(prediction2, fetch::math::function_tolerance<DataType>(),
  //                                  fetch::math::function_tolerance<DataType>()));
  //
  //  // train g
  //  layer.SetInput(label_name, labels);
  //  TypeParam loss = layer.Evaluate(error_output);
  //  layer.BackPropagateError(error_output);
  //  layer.Step(DataType{0.1f});
  //
  //  // train g2
  //  layer2.SetInput(label_name, labels);
  //  TypeParam loss2 = layer2.Evaluate(error_output);
  //  layer2.BackPropagateError(error_output);
  //  layer2.Step(DataType{0.1f});
  //
  //  EXPECT_TRUE(loss.AllClose(loss2, fetch::math::function_tolerance<DataType>(),
  //                            fetch::math::function_tolerance<DataType>()));
  //
  //  // new random input
  //  query_data.FillUniformRandom();
  //  key_data.FillUniformRandom();
  //  value_data.FillUniformRandom();
  //
  //  layer.SetInput("MultiheadAttention_Query", query_data);
  //  layer.SetInput("MultiheadAttention_Key", key_data);
  //  layer.SetInput("MultiheadAttention_Value", value_data);
  //  TypeParam prediction3 = layer.Evaluate(output_name);
  //
  //  layer2.SetInput("MultiheadAttention_Query", query_data);
  //  layer2.SetInput("MultiheadAttention_Key", key_data);
  //  layer2.SetInput("MultiheadAttention_Value", value_data);
  //  TypeParam prediction4 = layer2.Evaluate(output_name);
  //
  //  EXPECT_FALSE(prediction.AllClose(prediction3, fetch::math::function_tolerance<DataType>(),
  //                                   fetch::math::function_tolerance<DataType>()));
  //
  //  EXPECT_TRUE(prediction3.AllClose(prediction4, fetch::math::function_tolerance<DataType>(),
  //                                   fetch::math::function_tolerance<DataType>()));

**Comment by evsmithx** [comment]
Test for self attention

//  // rebuild
  //  using DataType  = typename TypeParam::Type;
  //  auto layer2 = *(fetch::ml::utilities::BuildLayer<TypeParam, LayerType>(dsp2));
  //
  //  // test equality
  //  layer.SetInput(input_name, input);
  //  prediction = layer.Evaluate(output_name, true);
  //  layer2.SetInput(input_name, input);
  //  TypeParam prediction2 = layer2.Evaluate(output_name, true);
  //
  //  ASSERT_TRUE(prediction.AllClose(prediction2, fetch::math::function_tolerance<DataType>(),
  //                                  fetch::math::function_tolerance<DataType>()));
  //
  //  // train g
  //  layer.SetInput(label_name, labels);
  //  TypeParam loss = layer.Evaluate(error_output);
  //  layer.BackPropagateError(error_output);
  //  layer.Step(DataType{0.1f});
  //
  //  // train g2
  //  layer2.SetInput(label_name, labels);
  //  TypeParam loss2 = layer2.Evaluate(error_output);
  //  layer2.BackPropagateError(error_output);
  //  layer2.Step(DataType{0.1f});
  //
  //  EXPECT_TRUE(loss.AllClose(loss2, fetch::math::function_tolerance<DataType>(),
  //                            fetch::math::function_tolerance<DataType>()));
  //
  //  // new random input
  //  input.FillUniformRandom();
  //
  //  layer.SetInput(input_name, input);
  //  TypeParam prediction3 = layer.Evaluate(output_name);
  //
  //  layer2.SetInput(input_name, input);
  //  TypeParam prediction4 = layer2.Evaluate(output_name);
  //
  //  EXPECT_FALSE(prediction.AllClose(prediction3, fetch::math::function_tolerance<DataType>(),
  //                                   fetch::math::function_tolerance<DataType>()));
  //
  //  EXPECT_TRUE(prediction3.AllClose(prediction4, fetch::math::function_tolerance<DataType>(),
  //                                   fetch::math::function_tolerance<DataType>()));
